src/lib/optimizer/strategy/dependent_group_by_reduction_rule.cpp

```diff
@@ -24,155 +24,103 @@ using namespace opossum;  // NOLINT
-
-    // Add the ANY() aggregate to the list of aggregate columns.
-    const auto aggregate_any_expression = any_(lqp_column_({stored_table_node, group_by_column}));
-    aggregate_node.node_expressions.emplace_back(aggregate_any_expression);
   }
 
   return group_by_list_changed;
 }
+
 }  // namespace
 
 namespace opossum {
 
-void DependentGroupByReductionRule::apply_to(const std::shared_ptr<AbstractLQPNode>& lqp) const {
-  // Store a copy of the root's column expressions.
-  const auto root_column_expressions = lqp->column_expressions();
-  visit_lqp(lqp, [&](const auto& node) {
+void DependentGroupByReductionRule::apply_to(const std::shared_ptr<AbstractLQPNode>& root_lqp) const {
+  visit_lqp(root_lqp, [&](const auto& node) {
     if (node->type != LQPNodeType::Aggregate) {
       return LQPVisitation::VisitInputs;
     }
     auto& aggregate_node = static_cast<AggregateNode&>(*node);
 
-    std::unordered_map<std::shared_ptr<const StoredTableNode>, std::set<ColumnID>> group_by_columns_per_table;
-    // Collect the group-by columns for each table in the aggregate node
-    for (auto expression_idx = size_t{0}; expression_idx < aggregate_node.aggregate_expressions_begin_idx;
-         ++expression_idx) {
-      const auto& expression = aggregate_node.node_expressions[expression_idx];
-
-      // Check that group by columns are not nullable. Unique columns can generally store NULLs while previous
-      // operators (e.g., outer joins) might have added NULLs to a primary key column. For now, we take the safe route
-      // and ignore all cases where group-by columns are nullable.
-      if (expression->is_nullable_on_lqp(aggregate_node)) {
-        continue;
-      }
-
-      const auto& column_expression = std::dynamic_pointer_cast<LQPColumnExpression>(expression);
-      if (!column_expression) {
-        // In case the group-by column is not a column expression (e.g., grouping by `a+1`), we take the safe route and
-        // continue even though constraints often hold for such simple expression.
-        continue;
-      }
-
-      const auto& stored_table_node =
-          std::dynamic_pointer_cast<const StoredTableNode>(column_expression->column_reference.original_node());
-      // If column is not a physical column skip
-      if (!stored_table_node) continue;
-
-      const auto column_id = column_expression->column_reference.original_column_id();
-      group_by_columns_per_table[stored_table_node].insert(column_id);
-    }
+    // Early exit: If there are no functional dependencies, we can skip this rule.
+    auto fds = aggregate_node.functional_dependencies();
+    if (fds.empty()) return LQPVisitation::VisitInputs;
 
-    // Store copy of the aggregate's column expressions to later check if this order is the query's final column order
+    // --- Preparation ---
+    // Store a copy of the root's column expressions before applying the rule
+    const auto root_column_expressions = root_lqp->column_expressions();
```

Comment:
This is now called more than once. Remember that column_expressions is somewhat expensive.

Reply:
In every iteration of `visit_lqp`

Reply:
But only for AggregateNodes since we got the following early exit:

```cpp
    // Early exit: If there are no functional dependencies, we can skip this rule.
    auto fds = aggregate_node.functional_dependencies();
    if (fds.empty()) return LQPVisitation::VisitInputs;
```

Reply:
Still less efficient than it could be, but not as bad as I thought